File: ruby/wordy/wordy.rb
class WordProblem
  OPS = {
    "plus" => :+,
    "minus" => :-,
    "multiplied by" => :*,
    "divided by" => :/
  }
  ANY_OP = OPS.keys.join('|')
  PHRASE_PATTERN = Regexp.new("^What is (?<memo>-?\\d+)(?<ops>(?: (?:#{ANY_OP}) -?\\d+)*)\\?$")
  OPS_PATTERN = Regexp.new(" (#{ANY_OP}) (-?\\d+)")

  def initialize(phrase)
    @phrase = phrase
  end

  def answer
    match = PHRASE_PATTERN.match(@phrase)
    raise ArgumentError, "Invalid problem phrase: #{@phrase}" if match.nil?
    value = match[:memo].to_i
    match[:ops].scan(OPS_PATTERN) do |op, new_value|
      value = value.__send__(OPS[op], new_value.to_i)
    end
    value
  end
end

File: javascript/wordy/wordy.js
export class ArgumentError extends Error {
    constructor(...params) {
        super(...params);
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, ArgumentError);
        }
        this.name = 'ArgumentError';
    }
}

export class WordProblem {
    constructor(question) {
        this.question = question;
    }

    answer() {
        const matches = /^What is (-?\d+)((?: (?:plus|minus|multiplied by|divided by) -?\d+)+)\?$/.exec(this.question);
        if (matches === null) {
            throw new ArgumentError('Wrong question format: ' + this.question);
        }

        let [, seed, ops] = matches;
        const opsRegex = / (plus|minus|multiplied by|divided by) (-?\d+)/y;
        let opMatch;
        let result = Number(seed);
        while ((opMatch = opsRegex.exec(ops)) !== null) {
            const [, op, num] = opMatch;
            if (op === 'plus') {
                result += Number(num);
            } else if (op === 'minus') {
                result -= Number(num);
            } else if (op === 'multiplied by') {
                result *= Number(num);
            } else if (op === 'divided by') {
                result /= Number(num);
            } else {
                throw new ArgumentError('Wrong operator: ' + op);
            }
        }
        return result;
    }
}
